File: tests/filter.test.ts
import { test, expect } from "bun:test";
import { createFilter } from "../src/filter";

test("deny rules override allow rules", () => {
  const filter = createFilter({
    allow: [".*"],
    deny: ["^git status$"]
  });

  expect(filter.matches("brew install gh")).toBe(true);
  expect(filter.matches("git status")).toBe(false);
});

test("allow list restricts commands", () => {
  const filter = createFilter({
    allow: ["^brew install", "^npm i"],
    deny: []
  });

  expect(filter.matches("brew install jq")).toBe(true);
  expect(filter.matches("npm i lodash")).toBe(true);
  expect(filter.matches("pip install requests")).toBe(false);
});

File: internal/filter/filter_test.go
package filter

import "testing"

func TestDenyOverridesAllow(t *testing.T) {
	f := New([]string{".*"}, []string{"^git status$"})
	if !f.Matches("brew install gh") {
		t.Fatal("expected allowed command")
	}
	if f.Matches("git status") {
		t.Fatal("expected denied command")
	}
}

func TestAllowRestricts(t *testing.T) {
	f := New([]string{"^brew install", "^npm i"}, nil)
	if !f.Matches("brew install jq") {
		t.Fatal("expected brew install to match")
	}
	if f.Matches("pip install requests") {
		t.Fatal("expected pip install not to match")
	}
}
